iterator.js

```diff
@@ -40,33 +59,56 @@ Iterator.prototype.createIterator = function() {
     keyRange: self._keyRange,
     autoContinue: false,
     order: self._order,
-    onError: function(err) { console.log('horrible error', err) },
+    onError: function(err) {
+      if (err.type !== 'abort' && !self._ended) {
+        // TODO: pass to next() callback
+        console.error('horrible error', err)
+      }
+    }
   })
 }
 
-// TODO the limit implementation here just ignores all reads after limit has been reached
-// it should cancel the iterator instead but I don't know how
 Iterator.prototype.onItem = function (value, cursor, cursorTransaction) {
-  if (!cursor && this.callback) {
-    this.callback()
-    this.callback = false
-    return
+  if (!cursor) {
+    this._finished = true
+  } else if (!!this._limit && this._limit > 0 && this._count++ >= this._limit) {
+    cursorTransaction.abort()
+    this._finished = true
+  } else {
+    this._cache.push(cursor.key, value)
+    cursor['continue']()
   }
-  var shouldCall = true
-
-  if (!!this._limit && this._limit > 0 && this._count++ >= this._limit)
-    shouldCall = false
 
-  if (shouldCall) this.callback(false, cursor.key, cursor.value)
-  if (cursor) cursor['continue']()
+  if (this._callback) {
+    this._next(this._callback)
+    this._callback = null
+  }
 }
 
+// TODO: use setImmediate (see memdown)
 Iterator.prototype._next = function (callback) {
+  // TODO: can remove this after upgrading abstract-leveldown
   if (!callback) return new Error('next() requires a callback argument')
```

Comment:
Returning an error looks a bit odd, shouldn't it throw?

Reply:
Woa, dunno why I missed that 😄